File: packages/shared/src/workflow.ts
/**
 * Workflow contract — Zod 4 schemas + inferred TypeScript types for the
 * canonical Janusly workflow shape.
 *
 * This module is the single source of truth for what a workflow, node, edge,
 * and metadata block look like. Every other workspace either parses incoming
 * payloads against `WorkflowSchema` or writes outgoing payloads matching it,
 * which is how the AI Studio, the engine runtime, and the persisted DAG JSON
 * stay in lockstep.
 *
 * Used by:
 * - `apps/api/src/routes/*` — `WorkflowSchema.parse` on `/start`, `/save`,
 *   `/validate`, AI generation, solution-pack import, and replay routes.
 * - `packages/engine/src/workflow-validation.ts` — graph-level validation runs
 *   on top of the schema-level parse.
 * - `packages/engine/src/start-run.ts` / `resume-run.ts` — read parsed nodes.
 * - `packages/engine/src/worker.ts` — `NodeSchema.parse(job.data)` on every
 *   BullMQ job (poison-payload protection).
 *
 * Invariants:
 * - `nodeTypeValues` is the closed set of node kinds the runtime supports.
 *   Adding a new kind requires both an entry here AND a matching executor in
 *   `packages/engine/src/node-registry.ts`.
 * - `dslVersion` defaults to "1.0"; bumping the version is a breaking change
 *   that requires a migration of persisted workflow JSON in `workflow_versions`.
 * - Records use Zod 4's two-arg `z.record(z.string(), z.unknown())` form (per
 *   AGENTS.md). Don't drop back to the single-arg legacy shape.
 */

import { z } from "zod";
import { WorkflowRecoverySchema } from "./recovery-contract";

/**
 * Current DSL version. Persisted workflows store this; older versions need a
 * migration before they parse cleanly against `WorkflowSchema`.
 */
export const workflowDslVersion = "1.0" as const;

/** Highest workflow version representable by PostgreSQL's `integer` column. */
export const workflowVersionMax = 2_147_483_647;

/**
 * Closed set of node-type discriminators. Adding a new value here is a
 * one-step compile-time check across the runtime — every consumer that
 * switches on `node.type` will fail to build until it handles the new case.
 */
export const nodeTypeValues = [
  "http",
  "condition",
  "tool",
  "agent",
  "multi_agent",
  "agent_reflection",
  "loop",
  "router",
  "router_llm",
  "transform",
  "ai",
  "webhook",
  "approval",
  "human_form",
  "noop",
  "subworkflow",
  "wait_until",
  "parallel_fork",
  "join",
  "schedule",
  "mcp_tool",
  // Event-driven trigger node types (config + inbound-payload contracts in
  // `trigger-types.ts`). Generic types use operator-promoted placeholders;
  // PagerDuty off-hours intent is compiled into a complete deterministic flow
  // by the prompt-generation route without expanding the LLM grammar.
  "webhook_received",
  "email_received",
  "file_dropped",
  "mcp_server_event",
  "pagerduty_incident",
] as const;

/** Zod enum derived from `nodeTypeValues`; powers `NodeSchema`'s `type` field. */
export const NodeTypeSchema = z.enum(nodeTypeValues);

/**
 * Single node in a workflow DAG.
 *
 * `config` is intentionally `Record<string, unknown>` — each node type
 * interprets its own config shape inside its executor. The runtime never
 * looks at `config` generically.
 */
export const NodeSchema = z.object({
  id: z.string().trim().min(1, "Node id is required"),
  type: NodeTypeSchema,
  /** Optional operator-authored display name. The runtime still dispatches by `type`. */
  label: z.string().trim().min(1).max(80).optional(),
  config: z.record(z.string(), z.unknown()).default({}),
});

/**
 * Directed edge between two nodes.
 *
 * `condition` is an optional Janusly expression (validated separately by
 * `packages/engine/src/expression.ts`); when present, the edge only fires when
 * the expression evaluates truthy.
 */
export const EdgeSchema = z.object({
  id: z.string().trim().min(1).optional(),
  from: z.string().trim().min(1, "Edge source is required"),
  to: z.string().trim().min(1, "Edge target is required"),
  condition: z.string().trim().min(1).optional(),
});

/**
 * Inline JSON-shaped descriptive metadata that lives on each workflow's
 * serialized form (`description`, `tags`). Distinct from the operational
 * metadata layer (owners + runbook + slackChannel etc.) that lives in
 * the `workflow_metadata` table and is keyed by `(orgId, workflowId)`.
 * `tags` defaults to `[]` so callers always see an array.
 */
export const WorkflowJsonMetadataSchema = z.object({
  description: z.string().trim().optional(),
  tags: z.array(z.string().trim().min(1)).default([]),
}).default({ tags: [] });

/**
 * Closed set of JSON-Schema-subset primitive `type` discriminators supported
 * by `WorkflowInputSchema`. Matches the small grammar
 * `packages/engine/src/inputs-validator.ts` understands — adding a new
 * primitive here means adding a matching branch there.
 */
export const workflowInputTypeValues = ["string", "number", "boolean", "object", "array"] as const;
/** Zod enum derived from `workflowInputTypeValues`. */
export const WorkflowInputTypeSchema = z.enum(workflowInputTypeValues);

/**
 * JSON-Schema-subset describing one input field on a workflow's declared
 * `inputs` shape. Recursive via `z.lazy` (object → properties → fields;
 * array → items). Powers `WorkflowSchema.inputs`.
 *
 * The grammar is intentionally limited:
 * - `type`: one of `workflowInputTypeValues`.
 * - `properties` + `required`: only meaningful for `type: "object"`.
 * - `items`: only meaningful for `type: "array"`.
 * - `enum`: closed set of literal values the field must equal one of.
 * - `description`: free-form, surfaces in the AI Studio Inspector.
 * - `default`: value used when the run-start payload omits the field.
 *
 * `default` is what makes a declared input a workflow-level SETTING rather
 * than only a per-run argument: the operator states the value once on the
 * workflow, every node reads it through `{{context.input.<name>}}`, and
 * changing it is one edit in one place instead of hunting literals across
 * node configs. The `{{inputs.*}}` scope is node-local inside a node config;
 * it refers to run input only in workflow-level `outputs` templates.
 * It is also what lets a trigger-driven workflow declare inputs at all — a
 * webhook/schedule run supplies `{ triggeredBy, event, … }`, never the
 * declared fields, so without defaults a required input would reject every
 * such run at `startRun`.
 *
 * No `oneOf` / `anyOf` / patterns / number ranges yet — expand the subset
 * (here AND in the hand-rolled validator) only when a real consumer needs it.
 */
export const WorkflowInputSchema: z.ZodType<WorkflowInputSchemaShape> = z.lazy(() => z.object({
  type: WorkflowInputTypeSchema,
  description: z.string().optional(),
  properties: z.record(z.string(), WorkflowInputSchema).optional(),
  required: z.array(z.string()).optional(),
  items: WorkflowInputSchema.optional(),
  enum: z.array(z.unknown()).optional(),
  default: z.unknown().optional(),
}));

/**
 * The "shape" type for a `WorkflowInputSchema` value. Hand-written rather than
 * `z.infer<>` because Zod can't derive the recursive shape through `z.lazy`
 * without an explicit type seed.
 */
export type WorkflowInputSchemaShape = {
  type: typeof workflowInputTypeValues[number];
  description?: string;
  properties?: Record<string, WorkflowInputSchemaShape>;
  required?: string[];
  items?: WorkflowInputSchemaShape;
  enum?: unknown[];
  default?: unknown;
};

/**
 * `outputs` is a record of output-name → template string. At terminal
 * (`succeeded`) status, the engine renders each template against the run's
 * context using `renderTemplate` (`packages/engine/src/template.ts`) and
 * persists the result to `runs.outputJson`. Shape:
 *   { result: "{{context.summarize.output.text}}", count: "{{context.fetch.output.length}}" }
 *
 * Outputs are intentionally string templates (not the same JSON-Schema
 * shape as `inputs`) because outputs are *projections* from existing
 * context, not types the workflow *declares* it produces. Reusing the
 * existing template substitution language avoids inventing a second one.
 */
export const WorkflowOutputsSchema = z.record(z.string(), z.string());

/** Editor-only coordinates persisted with the workflow, ignored by the runtime. */
export const WorkflowPositionSchema = z.object({
  x: z.number().finite(),
  y: z.number().finite(),
});

/**
 * Additive editor metadata. Keeping this separate from node execution fields
 * lets non-visual consumers ignore it while authoring clients restore layout.
 */
export const WorkflowUiSchema = z.object({
  positions: z.record(z.string().trim().min(1), WorkflowPositionSchema).optional(),
});

/**
 * Runtime handling for a node-config template whose path cannot be resolved.
 * `lenient` preserves the historical empty-string substitution while emitting
 * timeline evidence; `strict` fails the node before its executor can perform a
 * side effect. Optional on the wire so legacy snapshots remain byte-compatible.
 */
export const TemplatePolicySchema = z.enum(["lenient", "strict"]);

/**
 * Top-level workflow definition. Persisted as JSON in
 * `workflow_versions.dag_json` and consumed by the engine.
 */
export const WorkflowSchema = z.object({
  dslVersion: z.literal(workflowDslVersion).default(workflowDslVersion),
  id: z.string().trim().min(1).optional(),
  name: z.string().trim().min(1).optional(),
  metadata: WorkflowJsonMetadataSchema.optional(),
  inputs: WorkflowInputSchema.optional(),
  outputs: WorkflowOutputsSchema.optional(),
  templatePolicy: TemplatePolicySchema.optional(),
  recovery: WorkflowRecoverySchema.optional(),
  ui: WorkflowUiSchema.optional(),
  nodes: z.array(NodeSchema),
  edges: z.array(EdgeSchema),
}).superRefine((workflow, context) => {
  const positions = workflow.ui?.positions
  const nodeIds = new Set(workflow.nodes.map(node => node.id))
  if (positions) {
    for (const nodeId of Object.keys(positions)) {
      if (!nodeIds.has(nodeId)) {
        context.addIssue({
          code: "custom",
          path: ["ui", "positions", nodeId],
          message: "Workflow position must reference an existing node",
        });
      }
    }
  }
  for (const [index, effect] of (
    workflow.recovery?.contract?.effects ?? []
  ).entries()) {
    if (!nodeIds.has(effect.nodeId)) {
      context.addIssue({
        code: "custom",
        path: ["recovery", "contract", "effects", index, "nodeId"],
        message: "Recovery effect must reference an existing workflow node",
      });
    }
  }
});

/** Discriminator type for the closed set of node kinds. */
export type NodeType = z.infer<typeof NodeTypeSchema>;
/** Single workflow node, fully parsed (config defaulted to `{}`). */
export type WorkflowNode = z.infer<typeof NodeSchema>;
/** Single workflow edge, fully parsed. */
export type WorkflowEdge = z.infer<typeof EdgeSchema>;
/** Parsed metadata block with defaulted `tags`. */
export type WorkflowJsonMetadata = z.infer<typeof WorkflowJsonMetadataSchema>;
/** Closed set of JSON-Schema-subset primitive types. */
export type WorkflowInputType = z.infer<typeof WorkflowInputTypeSchema>;
/** Output-projection map (output-name → template string). */
export type WorkflowOutputs = z.infer<typeof WorkflowOutputsSchema>;
/** Runtime policy for unresolved node-config template paths. */
export type TemplatePolicy = z.infer<typeof TemplatePolicySchema>;
/** Persisted editor coordinates keyed by node id. */
export type WorkflowUi = z.infer<typeof WorkflowUiSchema>;
/** Top-level parsed workflow. */
export type Workflow = z.infer<typeof WorkflowSchema>;

File: internal/domain/workflow.go
// Package domain models the workflow document and ports the Janusly API's
// validation semantics: same issue codes, same messages, same check order,
// so a document rejected by one backend is rejected identically by the other.
// The porting source is the source contract and
// the source contract at the consistency pin.
package domain

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

// Node mirrors the wire shape of one workflow step. Config stays a loose map
// because each executor owns its own field contract.
type Node struct {
	ID     string         `json:"id"`
	Type   string         `json:"type"`
	Label  string         `json:"label,omitempty"`
	Config map[string]any `json:"config"`
}

// Edge mirrors the wire shape of one directed connection.
type Edge struct {
	ID        string `json:"id,omitempty"`
	From      string `json:"from"`
	To        string `json:"to"`
	Condition string `json:"condition,omitempty"`
	// OnError routes this edge only when the source node fails terminally
	// (retries exhausted). A failure with at least one on-error edge is
	// HANDLED: the run keeps going down the error branch instead of dying.
	OnError bool `json:"onError,omitempty"`
}

// WorkflowJSONMetadata is the small descriptive block persisted inside the
// versioned DAG. Operational ownership/runbook metadata lives in its own
// table and must never be smuggled into this document as arbitrary keys.
type WorkflowJSONMetadata struct {
	Description string   `json:"description,omitempty"`
	Tags        []string `json:"tags"`
}

// WorkflowPosition is editor-only state. The runtime ignores it, but save and
// load preserve finite coordinates for nodes that exist in the same DAG.
type WorkflowPosition struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type WorkflowUI struct {
	Positions map[string]WorkflowPosition `json:"positions,omitempty"`
}

// Workflow is the parsed document. Unknown top-level fields are ignored on
// parse, matching the contract schema's strip behavior.
type Workflow struct {
	DSLVersion     string                `json:"dslVersion"`
	ID             string                `json:"id,omitempty"`
	Name           string                `json:"name,omitempty"`
	Metadata       *WorkflowJSONMetadata `json:"metadata,omitempty"`
	Inputs         *InputSchema          `json:"inputs,omitempty"`
	Outputs        map[string]string     `json:"outputs,omitempty"`
	TemplatePolicy string                `json:"templatePolicy,omitempty"`
	Recovery       *WorkflowRecovery     `json:"recovery,omitempty"`
	UI             *WorkflowUI           `json:"ui,omitempty"`
	Nodes          []Node                `json:"nodes"`
	Edges          []Edge                `json:"edges"`
}

const dslVersion = "1.0"

// rawWorkflow separates decoding from contract checking so a malformed field
// reports its path instead of failing the whole document opaquely.
type rawWorkflow struct {
	DSLVersion     *string           `json:"dslVersion"`
	ID             *string           `json:"id"`
	Name           *string           `json:"name"`
	Inputs         *InputSchema      `json:"inputs"`
	Outputs        map[string]string `json:"outputs"`
	TemplatePolicy *string           `json:"templatePolicy"`
	Recovery       *WorkflowRecovery `json:"recovery"`
	Nodes          *[]rawNode        `json:"nodes"`
	Edges          *[]rawEdge        `json:"edges"`
}

type rawNode struct {
	ID     *string        `json:"id"`
	Type   *string        `json:"type"`
	Label  *string        `json:"label"`
	Config map[string]any `json:"config"`
}

type rawEdge struct {
	ID        *string `json:"id"`
	From      *string `json:"from"`
	To        *string `json:"to"`
	Condition *string `json:"condition"`
	OnError   *bool   `json:"onError"`
}

// Parse decodes and contract-checks a workflow document. Contract violations
// come back as `invalid_contract` issues carrying the field path — the same
// code the contract emits when its schema parse fails — and no workflow is
// returned alongside them. Message wording differs from the contract's
// schema library; code and path consistency is the contract.
func Parse(raw []byte) (*Workflow, []Issue) {
	var doc rawWorkflow
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, []Issue{{Code: CodeInvalidContract, Message: "workflow: " + err.Error()}}
	}
	var document map[string]json.RawMessage
	if err := json.Unmarshal(raw, &document); err != nil || document == nil {
		return nil, []Issue{{Code: CodeInvalidContract, Message: "workflow: expected an object"}}
	}

	var issues []Issue
	contract := func(path, message string) {
		issues = append(issues, Issue{Code: CodeInvalidContract, Message: path + ": " + message})
	}

	wf := &Workflow{
		DSLVersion: dslVersion,
		Metadata:   &WorkflowJSONMetadata{Tags: []string{}},
		Outputs:    doc.Outputs,
		Inputs:     doc.Inputs,
	}
	for _, field := range []string{"dslVersion", "id", "name", "metadata", "inputs", "outputs", "templatePolicy", "recovery", "ui"} {
		if value, present := document[field]; present && isJSONNull(value) {
			contract(field, "expected a non-null value")
		}
	}
	if encoded, present := document["metadata"]; present && !isJSONNull(encoded) {
		metadata, problems := parseWorkflowJSONMetadata(encoded)
		if len(problems) == 0 {
			wf.Metadata = metadata
		} else {
			for _, problem := range problems {
				contract("metadata"+problem.path, problem.message)
			}
		}
	}
	if doc.Inputs != nil && (!validInputSchemaWire(document["inputs"]) || !validInputSchemaShape(doc.Inputs)) {
		contract("inputs", fmt.Sprintf("expected a supported recursive schema with at most %d nodes", InputSchemaNodeMax))
	}
	if doc.Outputs != nil && !validWorkflowOutputsWire(document["outputs"]) {
		contract("outputs", "expected an object of string templates")
	}
	if doc.DSLVersion != nil && strings.TrimSpace(*doc.DSLVersion) != dslVersion {
		contract("dslVersion", fmt.Sprintf("expected %q", dslVersion))
	}
	if doc.ID != nil {
		wf.ID = strings.TrimSpace(*doc.ID)
		if wf.ID == "" {
			contract("id", "expected a non-empty string")
		}
	}
	if doc.Name != nil {
		wf.Name = strings.TrimSpace(*doc.Name)
		if wf.Name == "" {
			contract("name", "expected a non-empty string")
		}
	}
	if doc.TemplatePolicy != nil {
		wf.TemplatePolicy = *doc.TemplatePolicy
		if wf.TemplatePolicy != "lenient" && wf.TemplatePolicy != "strict" {
			contract("templatePolicy", `expected "lenient" or "strict"`)
		}
	}
	if doc.Recovery != nil {
		wf.Recovery = doc.Recovery
		for _, problem := range validateWorkflowRecoveryWire(document["recovery"]) {
			contract("recovery"+problem.path, problem.message)
		}
		normalizeRecoveryContract(doc.Recovery.Contract)
		// The versioned contract validates at parse time — the contract
		// rejects an invalid contract in WorkflowSchema.parse, so the same
		// document fails here with path-prefixed invalid_contract issues.
		for _, problem := range ValidateRecoveryContract(doc.Recovery.Contract) {
			contract("recovery.contract", problem)
		}
		if !isJSONNull(doc.Recovery.CircuitBreaker) {
			if _, _, problem := ParseCircuitBreakerThreshold(doc.Recovery.CircuitBreaker); problem != "" {
				contract("recovery", problem)
			}
		}
	}

	if doc.Nodes == nil {
		contract("nodes", "Invalid input: expected array, received undefined")
	} else {
		// The raw pass only tells an explicit null from an absent field. A
		// document without the token cannot carry one, so the stored
		// snapshot re-parsed on every claim decodes its nodes once.
		var rawNodes []map[string]json.RawMessage
		if bytes.Contains(document["nodes"], []byte("null")) {
			_ = json.Unmarshal(document["nodes"], &rawNodes)
		}
		// Non-nil even when empty: the workflow snapshot persisted at run
		// start must round-trip as "nodes": [] — a nil slice would marshal
		// as null and fail this same contract on re-parse.
		wf.Nodes = []Node{}
		for i, n := range *doc.Nodes {
			path := "nodes." + strconv.Itoa(i)
			if i < len(rawNodes) && rawNodes[i] != nil {
				for _, field := range []string{"label", "config"} {
					if value, present := rawNodes[i][field]; present && isJSONNull(value) {
						contract(path+"."+field, "expected a non-null value")
					}
				}
			}
			node := Node{Config: n.Config}
			if node.Config == nil {
				node.Config = map[string]any{}
			}
			if n.ID == nil || strings.TrimSpace(*n.ID) == "" {
				contract(path+".id", "Node id is required")
			} else {
				node.ID = strings.TrimSpace(*n.ID)
			}
			if n.Type == nil || strings.TrimSpace(*n.Type) == "" {
				contract(path+".type", "Node type is required")
			} else {
				node.Type = strings.TrimSpace(*n.Type)
			}
			if n.Label != nil {
				node.Label = strings.TrimSpace(*n.Label)
				if node.Label == "" || len(node.Label) > 80 {
					contract(path+".label", "expected 1..80 characters")
				}
			}
			wf.Nodes = append(wf.Nodes, node)
		}
	}

	if doc.Edges != nil {
		wf.Edges = []Edge{}
	}
	if doc.Edges == nil {
		contract("edges", "Invalid input: expected array, received undefined")
	} else {
		var rawEdges []map[string]json.RawMessage
		if bytes.Contains(document["edges"], []byte("null")) {
			_ = json.Unmarshal(document["edges"], &rawEdges)
		}
		for i, e := range *doc.Edges {
			path := "edges." + strconv.Itoa(i)
			if i < len(rawEdges) && rawEdges[i] != nil {
				for _, field := range []string{"id", "condition", "onError"} {
					if value, present := rawEdges[i][field]; present && isJSONNull(value) {
						contract(path+"."+field, "expected a non-null value")
					}
				}
			}
			edge := Edge{}
			if e.ID != nil {
				edge.ID = strings.TrimSpace(*e.ID)
				if edge.ID == "" {
					contract(path+".id", "expected a non-empty string")
				}
			}
			if e.From == nil || strings.TrimSpace(*e.From) == "" {
				contract(path+".from", "Edge source is required")
			} else {
				edge.From = strings.TrimSpace(*e.From)
			}
			if e.To == nil || strings.TrimSpace(*e.To) == "" {
				contract(path+".to", "Edge target is required")
			} else {
				edge.To = strings.TrimSpace(*e.To)
			}
			if e.Condition != nil {
				edge.Condition = strings.TrimSpace(*e.Condition)
				if edge.Condition == "" {
					contract(path+".condition", "expected a non-empty string")
				}
			}
			if e.OnError != nil {
				edge.OnError = *e.OnError
			}
			wf.Edges = append(wf.Edges, edge)
		}
	}

	if encoded, present := document["ui"]; present && !isJSONNull(encoded) {
		nodeIDs := make(map[string]struct{}, len(wf.Nodes))
		for _, node := range wf.Nodes {
			nodeIDs[node.ID] = struct{}{}
		}
		ui, problems := parseWorkflowUI(encoded, nodeIDs)
		if len(problems) == 0 {
			wf.UI = ui
		} else {
			for _, problem := range problems {
				contract("ui"+problem.path, problem.message)
			}
		}
	}

	if len(issues) > 0 {
		return nil, issues
	}
	return wf, nil
}

type workflowFieldProblem struct {
	path    string
	message string
}

func parseWorkflowJSONMetadata(raw json.RawMessage) (*WorkflowJSONMetadata, []workflowFieldProblem) {
	var object map[string]json.RawMessage
	if json.Unmarshal(raw, &object) != nil || object == nil {
		return nil, []workflowFieldProblem{{message: "expected an object"}}
	}
	metadata := &WorkflowJSONMetadata{Tags: []string{}}
	var problems []workflowFieldProblem
	if value, present := object["description"]; present {
		if isJSONNull(value) || json.Unmarshal(value, &metadata.Description) != nil {
			problems = append(problems, workflowFieldProblem{path: ".description", message: "expected a string"})
		} else {
			metadata.Description = strings.TrimSpace(metadata.Description)
		}
	}
	if value, present := object["tags"]; present {
		if isJSONNull(value) || json.Unmarshal(value, &metadata.Tags) != nil || metadata.Tags == nil {
			problems = append(problems, workflowFieldProblem{path: ".tags", message: "expected an array of strings"})
		} else {
			for index := range metadata.Tags {
				metadata.Tags[index] = strings.TrimSpace(metadata.Tags[index])
				if metadata.Tags[index] == "" {
					problems = append(problems, workflowFieldProblem{
						path: fmt.Sprintf(".tags.%d", index), message: "expected a non-empty string",
					})
				}
			}
		}
	}
	return metadata, problems
}

func validWorkflowOutputsWire(raw json.RawMessage) bool {
	var outputs map[string]json.RawMessage
	if len(raw) == 0 || isJSONNull(raw) || json.Unmarshal(raw, &outputs) != nil || outputs == nil {
		return false
	}
	for _, encoded := range outputs {
		var template string
		if isJSONNull(encoded) || json.Unmarshal(encoded, &template) != nil {
			return false
		}
	}
	return true
}

func parseWorkflowUI(raw json.RawMessage, nodeIDs map[string]struct{}) (*WorkflowUI, []workflowFieldProblem) {
	var object map[string]json.RawMessage
	if json.Unmarshal(raw, &object) != nil || object == nil {
		return nil, []workflowFieldProblem{{message: "expected an object"}}
	}
	ui := &WorkflowUI{}
	encoded, present := object["positions"]
	if !present {
		return ui, nil
	}
	var positions map[string]map[string]json.RawMessage
	if isJSONNull(encoded) || json.Unmarshal(encoded, &positions) != nil || positions == nil {
		return nil, []workflowFieldProblem{{path: ".positions", message: "expected an object"}}
	}
	ui.Positions = make(map[string]WorkflowPosition, len(positions))
	var problems []workflowFieldProblem
	positionIDs := make([]string, 0, len(positions))
	for rawID := range positions {
		positionIDs = append(positionIDs, rawID)
	}
	sort.Strings(positionIDs)
	for _, rawID := range positionIDs {
		rawPosition := positions[rawID]
		id := strings.TrimSpace(rawID)
		path := ".positions." + rawID
		if id == "" {
			problems = append(problems, workflowFieldProblem{path: path, message: "expected a non-empty node id"})
			continue
		}
		if _, duplicate := ui.Positions[id]; duplicate {
			problems = append(problems, workflowFieldProblem{path: path, message: "duplicate node id after trimming"})
			continue
		}
		if _, exists := nodeIDs[id]; !exists {
			problems = append(problems, workflowFieldProblem{path: path, message: "position must reference an existing node"})
			continue
		}
		if rawPosition == nil {
			problems = append(problems, workflowFieldProblem{path: path, message: "expected an object"})
			continue
		}
		position := WorkflowPosition{}
		valid := true
		for _, axis := range []struct {
			name  string
			value *float64
		}{{"x", &position.X}, {"y", &position.Y}} {
			value, exists := rawPosition[axis.name]
			if !exists || isJSONNull(value) || json.Unmarshal(value, axis.value) != nil ||
				math.IsNaN(*axis.value) || math.IsInf(*axis.value, 0) {
				problems = append(problems, workflowFieldProblem{
					path: path + "." + axis.name, message: "expected a finite number",
				})
				valid = false
			}
		}
		if valid {
			ui.Positions[id] = position
		}
	}
	return ui, problems
}

// CanonicalWorkflowDocument serializes only the public workflow contract. It
// deliberately strips save-only carriers and unknown fields, applies parser
// defaults, and writes generated id/name values supplied by the caller.
func CanonicalWorkflowDocument(wf *Workflow) ([]byte, error) {
	if wf == nil {
		return nil, fmt.Errorf("workflow is nil")
	}
	metadata := wf.Metadata
	if metadata == nil {
		metadata = &WorkflowJSONMetadata{Tags: []string{}}
	}
	if metadata.Tags == nil {
		metadata = &WorkflowJSONMetadata{Description: metadata.Description, Tags: []string{}}
	}
	document := map[string]any{
		"dslVersion": wf.DSLVersion,
		"metadata":   metadata,
		"nodes":      wf.Nodes,
		"edges":      wf.Edges,
	}
	if wf.ID != "" {
		document["id"] = wf.ID
	}
	if wf.Name != "" {
		document["name"] = wf.Name
	}
	if wf.Inputs != nil {
		document["inputs"] = wf.Inputs
	}
	if wf.Outputs != nil {
		document["outputs"] = wf.Outputs
	}
	if wf.TemplatePolicy != "" {
		document["templatePolicy"] = wf.TemplatePolicy
	}
	if wf.Recovery != nil {
		document["recovery"] = wf.Recovery
	}
	if wf.UI != nil {
		document["ui"] = wf.UI
	}
	return json.Marshal(document)
}
